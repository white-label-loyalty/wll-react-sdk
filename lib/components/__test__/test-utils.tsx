import { render } from '@testing-library/react';
import React from 'react';
import { CoreApiUrl, WllSdkProvider } from '../../context/WllSdkContext';
import { defaultTheme } from '../../utils/styling';

const defaultConfig = {
  apiKey: 'test-api-key',
  userToken: 'test-user-token',
  coreApiUrl: CoreApiUrl.STAGING_EU,
};

const AllTheProviders = ({ children }: { children: React.ReactNode }) => {
  return (
    <WllSdkProvider
      config={defaultConfig}
      theme={defaultTheme}
      navigationConfig={{}}
    >
      {children}
    </WllSdkProvider>
  );
};

const customRender = (ui: React.ReactElement, options = {}) =>
  render(ui, { wrapper: AllTheProviders, ...options });

// re-export everything
export * from '@testing-library/react';

// override render method
export { customRender as render };
