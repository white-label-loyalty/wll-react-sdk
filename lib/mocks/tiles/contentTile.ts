import {
  ContentTileConfig,
  Tile,
  TileHeight,
  TileType,
} from '../../types/tile';

export type ContentTileMockConfig = Partial<ContentTileConfig> & {
  tileHeight?: TileHeight;
  active?: boolean;
};

const defaultConfig: ContentTileConfig = {
  title: 'Gold Tier Unlocked!',
  body: "You've unlocked exclusive Gold member benefits including 2X points on every purchase, priority support, and VIP event access.",
  artworkUrl:
    'https://images.pexels.com/photos/352097/pexels-photo-352097.jpeg',
};

export const createContentTileMock = (config?: ContentTileMockConfig): Tile => {
  const {
    tileHeight = TileHeight.Full,
    active = true,
    ...tileConfig
  } = config || {};

  const mergedConfig = { ...defaultConfig, ...tileConfig };

  return {
    id: '1',
    type: TileType.Content,
    active,
    createdAt: '',
    updatedAt: '',
    tileHeight,
    priority: 1,
    configuration: mergedConfig,
  };
};
