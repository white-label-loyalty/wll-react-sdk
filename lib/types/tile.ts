export enum CtaAction {
  sameWindow = 'SAME_WINDOW',
  newWindow = 'NEW_WINDOW',
}

export enum BadgeTileType {
  latestEarned = 'LATEST_EARNED',
  specificBadge = 'SPECIFIC_BADGE',
}

export enum TierTileType {
  currentTier = 'CURRENT_TIER',
  nextTier = 'NEXT_TIER',
  specificTier = 'SPECIFIC_TIER',
}

export enum TileType {
  Banner = 'BANNER',
  Points = 'POINTS',
  Content = 'CONTENT',
  Reward = 'REWARD',
  Badge = 'BADGE',
  RewardCategory = 'REWARD_CATEGORY',
  Tier = 'TIER',
}

export enum TileHeight {
  Half = 'HALF',
  Full = 'FULL',
}
type Badge = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  createdAt: string | null;
  updatedAt: string | null;
};
export class RewardCategory {
  name?: string | null;
  priority?: number;
  type?: string | null;
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  description?: string | null;
  metadata?: string | null;
  pictureUrl?: string | null;
  rewards?: Reward[];
  parent?: string | null;
}

export class Reward {
  id?: string | null;
  createdAt?: string | null;
  updatedAt?: string | null;
  name?: string | null;
  pictureUrl?: string | null;
  description?: string | null;
  value?: number | null;
  price?: number;
  priority?: number;
  discounts?: any[] | null;
  redemptionChannels?: string[] | null;
  venues?: any[] | null;
  purchasableForAudiences?: string[] | null;
  category?: string | null;
  terms?: string | null;
  availability?: Availability;
  purchasable?: boolean;
  tier?: string | null;
  summary?: string | null;
  redemptionMessage?: string | null;
  visibilityCriteria?: string | null;
  logoUrl?: string | null;
  type?: string | null;
  codeType?: string | null;
  code?: string | null;
  purchaseExpiration?: string | null;
  hideCode?: boolean;
  notificationConfig?: string | null;
}
type Availability = {
  start: string | null;
  end: string | null;
};
type Effectivity = {
  start: string | null;
  end: string | null;
};

export type TierType = {
  id: string | null;
  name: string | null;
  description: string | null;
  artworkUrl: string | null;
  priority: number;
  pointsRequirement: number;
  calculation: string | null;
  accumulationPeriod: string | null;
  effectivity: Effectivity;
  createdAt: string | null;
  updatedAt: string | null;
  earnedPoints: number;
  attained: boolean;
};

export class BannerTileConfig {
  imageUrl?: string | null;
  title?: string | null;
  description?: string | null;
  ctaText?: string | null;
  ctaLink?: string | null;
  ctaAction?: CtaAction;
}

export class PointsTileConfig {
  title?: string | null;
  multiplier?: number;
  prefix?: string | null;
  suffix?: string | null;
  imageUrl?: string | null;
  points?: number;
}

export class ContentTileConfig {
  title?: string | null;
  subtitle?: string | null;
  imageUrl?: string | null;
}

export class RewardTileConfig {
  reward?: Reward;
  showPrice?: boolean;
}

export class BadgeTileConfig {
  badgeTileType?: BadgeTileType;
  badgeId?: string | null;
  badge?: Badge;
}

export class RewardCategoryTileConfig {
  categoryId?: string | null;
  allowDecorationOverlay?: boolean;
  rewardCategory?: RewardCategory;
}

export class TierTileConfig {
  tierId?: string | null;
  tier?: TierType;
  type?: TierTileType;
}

export type Tile = {
  id: string | null;
  type: TileType;
  active: boolean;
  createdAt: string | null;
  updatedAt: string | null;
  visibilityCriteria: object | null;
  tileHeight: TileHeight;
  configuration: TileConfig;
};

const getConfigForTileType = (tileType: TileType) => {
  switch (tileType) {
    case TileType.Banner:
      return BannerTileConfig;
    case TileType.Points:
      return PointsTileConfig;
    case TileType.Content:
      return ContentTileConfig;
    case TileType.Reward:
      return RewardTileConfig;
    case TileType.Badge:
      return BadgeTileConfig;
    case TileType.RewardCategory:
      return RewardCategoryTileConfig;
    case TileType.Tier:
      return TierTileConfig;
    default:
      throw new Error(`Tile type ${tileType} is not supported`);
  }
};

export type TileConfig = InstanceType<
  ReturnType<typeof getConfigForTileType>
>;
