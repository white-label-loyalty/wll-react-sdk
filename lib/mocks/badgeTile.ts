import { BadgeTileType, TileHeight, TileType } from '../types/tile';

export type BadgeTileMockConfig = {
  id?: string;
  active?: boolean;
  tileHeight?: TileHeight;
  type?: BadgeTileType;
  badgeId?: string;
  internalName?: string;
  name?: string;
  description?: string;
  artworkUrl?: string;
  emptyBadgeArtworkUrl?: string;
  badgeNotEarnedMessage?: string;
  count?: number;
};

export const createBadgeTileMock = (config: BadgeTileMockConfig = {}) => {
  const {
    id = 'badge-tile',
    active = true,
    tileHeight = TileHeight.Full,
    type = BadgeTileType.Latest,
    badgeId = '900a2477-95c4-4c42-ae2d-3795e7f0f5f2',
    internalName = 'Top Spender',
    name = 'Top Spender',
    description = 'Spent Â£100 on 5 Separate transactions',
    artworkUrl = 'https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/',
    emptyBadgeArtworkUrl = 'https://ucarecdn.com/3d3731b2-faec-4779-9cd8-3691631d280c/',
    badgeNotEarnedMessage,
    count = 0,
  } = config;

  return {
    id,
    type: TileType.Badge,
    active,
    createdAt: '2024-08-06T08:53:24.307Z',
    updatedAt: '2024-08-06T08:53:24.307Z',
    tileHeight,
    priority: 0,
    configuration: {
      type,
      badgeId,
      internalName,
      name,
      description,
      artworkUrl,
      priority: 0,
      status: 'ACTIVE',
      createdAt: '2024-08-06T08:53:24.307Z',
      updatedAt: '2024-08-06T08:53:24.307Z',
      awardedDatePrefix: 'Awarded',
      emptyBadgeMessage: "You haven't earned any badges yet.",
      badgeNotEarnedMessage:
        'badgeNotEarnedMessage' in config
          ? badgeNotEarnedMessage
          : 'Badge not earned yet',
      emptyBadgeArtworkUrl,
      count,
    },
  };
};
