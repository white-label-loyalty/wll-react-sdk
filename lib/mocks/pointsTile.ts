import { TileHeight, TileType } from '../types/tile';

export type PointsTileConfig = {
  title: string;
  points: number;
  pointsPrefix?: string;
  pointsSuffix?: string;
  artworkUrl: string;
  pointsMultiplier?: number;
};

export type PointsTileMockConfig = Partial<PointsTileConfig> & {
  active?: boolean;
  tileHeight?: TileHeight;
};

export const createPointsTileMock = (config?: PointsTileMockConfig) => {
  const {
    active = true,
    tileHeight = TileHeight.Full,
    title = 'Available Points',
    points = 1000,
    pointsPrefix = '',
    pointsSuffix = ' pts',
    artworkUrl = 'https://example.com/points.png',
    pointsMultiplier,
  } = config || {};

  return {
    id: '1',
    type: TileType.Points,
    active,
    createdAt: '2025-01-26T11:13:38Z',
    updatedAt: '2025-01-26T11:13:38Z',
    tileHeight,
    priority: 1,
    configuration: {
      title,
      points,
      pointsPrefix,
      pointsSuffix,
      artworkUrl,
      ...(pointsMultiplier !== undefined && { pointsMultiplier }),
    },
  };
};
