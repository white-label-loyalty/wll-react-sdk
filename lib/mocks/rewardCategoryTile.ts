import { RewardCategoryTileConfig, TileHeight, TileType } from '../types/tile';

export type RewardCategoryTileMockConfig = Partial<RewardCategoryTileConfig> & {
  active?: boolean;
  tileHeight?: TileHeight;
};

export const createRewardCategoryTileMock = (
  config?: RewardCategoryTileMockConfig
) => {
  const defaultConfig = {
    active: true,
    tileHeight: TileHeight.Full,
    showName: true,
    rewardCategoryId: '9d8287e0-bae9-47b4-8bf0-0115fd64336a',
    artworkUrl:
      'https://images.pexels.com/photos/1850595/pexels-photo-1850595.jpeg?auto=compress&cs=tinysrgb&w=1260&h=750&dpr=2',
    name: 'Restaurant Rewards',
  };

  const { active, tileHeight, showName, rewardCategoryId, artworkUrl, name } = {
    ...defaultConfig,
    ...config,
  };

  return {
    id: 'f57f2f98-cabf-4b81-82ac-a0249172d65c',
    type: TileType.RewardCategory,
    active,
    createdAt: '2024-10-18T14:40:37.835Z',
    updatedAt: '2024-10-18T14:40:37.835Z',
    tileHeight,
    priority: 8,
    configuration: {
      showName,
      rewardCategoryId,
      artworkUrl,
      name,
    },
  };
};
