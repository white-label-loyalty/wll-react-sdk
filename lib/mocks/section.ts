import { SectionType, TSection } from '../types/section';
import { Tile, TileType } from '../types/tile';
import { createTileMock } from './tiles';

export type SectionMockConfig = Partial<TSection> & {
  tiles?: Tile[];
};

const defaultConfig: TSection = {
  details: [],
  defaultLocale: 'en',
  name: 'Elite Member Benefits',
  type: SectionType.Grid,
  active: true,
  id: 'a35399a7-580a-4493-afff-24ef4facf3f4',
  createdAt: '2024-11-19T15:50:37.701Z',
  updatedAt: '2024-12-06T11:18:16.842Z',
  priority: 2,
  tiles: [createTileMock(TileType.Content)],
  title: 'Elite Member Benefits',
  locale: 'en',
  pointsSuffix: 'miles',
};

export const createSectionMock = (config?: SectionMockConfig): TSection => {
  const { tiles, ...sectionConfig } = config || {};
  
  // If tiles are provided in config, use those. Otherwise use default tiles
  const mergedTiles = tiles || defaultConfig.tiles;

  return {
    ...defaultConfig,
    ...sectionConfig,
    tiles: mergedTiles,
  };
};

// Helper function to create a section with specific tile types
export const createSectionWithTiles = (tiles: Tile[]): TSection => {
  return createSectionMock({ tiles });
};
