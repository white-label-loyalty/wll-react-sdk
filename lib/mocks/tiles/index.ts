import { Tile, TileType } from '../../types/tile';
import { createBadgeTileMock } from './badgeTile';
import { createBannerTileMock } from './bannerTile';
import { createContentTileMock } from './contentTile';
import { createPointsTileMock } from './pointsTile';
import { createRewardCategoryTileMock } from './rewardCategoryTile';
import { createRewardTileMock } from './rewardTile';
export { createContentTileMock } from './contentTile';

export const createTileMock = (type: TileType, config?: any): Tile => {
  switch (type) {
    case TileType.Content:
      return createContentTileMock(config);
    case TileType.Reward:
      return createRewardTileMock(config);
    case TileType.Banner:
      return createBannerTileMock(config);
    case TileType.Points:
      return createPointsTileMock(config);
    case TileType.Badge:
      return createBadgeTileMock(config);
    case TileType.RewardCategory:
      return createRewardCategoryTileMock(config);
    case TileType.Tier:
    // TODO: Implement other tile types as they are needed
    default:
      throw new Error(`Tile type ${type} not yet implemented`);
  }
};
