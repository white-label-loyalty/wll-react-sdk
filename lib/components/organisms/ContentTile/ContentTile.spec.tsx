import React from 'react';
import { createContentTileMock } from '../../../mocks/contentTile';
import { TileHeight } from '../../../types/tile';
import { render } from '../../__test__/test-utils';
import { ContentTile } from './index';

const ContentTileMock = createContentTileMock();

describe('<ContentTile />', () => {
  it('renders and matches snapshot', () => {
    const { container } = render(<ContentTile tile={ContentTileMock} />);
    expect(container).toMatchSnapshot();
  });

  it('handles inactive tile gracefully', () => {
    const { container } = render(
      <ContentTile tile={{ ...ContentTileMock, active: false }} />
    );

    expect(container.firstChild).toBeNull();
  });
});

describe('<ContentTile /> Rendering States', () => {
  it('if full-size tile has no artwork, content should be rendered', () => {
    const fullSizeTileWithNoArtwork = createContentTileMock({
      tileHeight: TileHeight.Full,
      artworkUrl: undefined,
    });
    const { queryByTestId } = render(
      <ContentTile tile={fullSizeTileWithNoArtwork} />
    );
    expect(queryByTestId('tile-content')).toBeTruthy();
  });

  it('if half-size tile has artwork, content should be null', () => {
    const halfSizeTileWithArtwork = createContentTileMock({
      tileHeight: TileHeight.Half,
      artworkUrl:
        'https://images.pexels.com/photos/352097/pexels-photo-352097.jpeg',
    });
    const { queryByTestId } = render(
      <ContentTile tile={halfSizeTileWithArtwork} />
    );
    expect(queryByTestId('tile-content')).toBeNull();
  });

  it('handles inactive tile gracefully', () => {
    const inactiveTile = createContentTileMock({
      active: false,
    });

    const { container } = render(<ContentTile tile={inactiveTile} />);
    expect(container.firstChild).toBeNull();
  });
});
