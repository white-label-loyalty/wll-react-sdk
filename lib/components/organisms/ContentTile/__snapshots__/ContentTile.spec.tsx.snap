// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ContentTile /> renders and matches snapshot 1`] = `
<div>
  <button
    aria-disabled="true"
    aria-label="Gold Tier Unlocked!"
    class="css-view-175oi2r r-pointerEvents-12vffkv r-aspectRatio-1ujkv8a r-borderRadius-y47klf r-height-1pi2tsx r-overflow-1udh08x r-position-bnwqim r-width-13qz1uu"
    disabled=""
    role="button"
    style="background-color: rgb(248, 247, 252); opacity: 1; flex-direction: column; justify-content: flex-start; align-items: stretch;"
    tabindex="-1"
    type="button"
  >
    <button
      aria-disabled="true"
      aria-label="Gold Tier Unlocked!"
      class="css-view-175oi2r r-pointerEvents-12vffkv r-aspectRatio-1ujkv8a r-borderRadius-y47klf r-height-1pi2tsx r-overflow-1udh08x r-position-bnwqim r-width-13qz1uu"
      disabled=""
      role="button"
      style="background-color: rgb(248, 247, 252); opacity: 1; flex-direction: column; justify-content: flex-start; align-items: stretch;"
      tabindex="-1"
      type="button"
    >
      <div
        class="css-view-175oi2r r-alignItems-1awozwy r-height-1pi2tsx r-justifyContent-1777fci r-position-bnwqim r-flexBasis-1t2qqvi r-objectFit-1t7uo4s r-width-13qz1uu"
        style="background-color: rgba(57, 46, 215, 0.2);"
      >
        <div
          class="css-view-175oi2r r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          style="background-color: rgba(57, 46, 215, 0.2); opacity: 1;"
        />
        <div
          aria-label="Content image for Gold Tier Unlocked!"
          class="css-view-175oi2r r-flexBasis-1mlwlqe r-overflow-1udh08x r-zIndex-417010 r-bottom-1p0dtai r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af"
          data-testid="tile-media"
          style="opacity: 0;"
        >
          <div
            class="css-view-175oi2r r-backgroundColor-1niwhzg r-backgroundPosition-vvn4in r-backgroundRepeat-u6sd8q r-bottom-1p0dtai r-height-1pi2tsx r-left-1d2f490 r-position-u8s1d r-right-zchlnj r-top-ipm5af r-width-13qz1uu r-zIndex-1wyyakw r-backgroundSize-4gszlv"
            style="background-image: url(https://images.pexels.com/photos/352097/pexels-photo-352097.jpeg);"
          />
          <img
            alt="Content image for Gold Tier Unlocked!"
            class="css-accessibilityImage-9pa8cd"
            draggable="false"
            src="https://images.pexels.com/photos/352097/pexels-photo-352097.jpeg"
          />
        </div>
      </div>
      <div
        class="css-view-175oi2r r-display-6koalj"
        data-testid="tile-content"
        style="justify-content: center;"
      >
        <div
          class="css-view-175oi2r r-flex-13awgt0 r-justifyContent-1777fci r-paddingInline-3o4zer"
        >
          <div
            class="css-view-175oi2r r-alignItems-1awozwy r-flexDirection-18u37iz r-justifyContent-1wtj0ep r-marginBottom-5oul0u"
            data-testid="tile-header"
            style="margin-top: 12px;"
          >
            <div
              aria-label="Gold Tier Unlocked!"
              class="css-text-146c3p1 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01 r-whiteSpace-3s2u2q r-wordWrap-1iln25a"
              data-testid="tile-title"
              dir="auto"
              style="color: rgb(0, 0, 0); font-size: 24px; font-weight: bold;"
            >
              Gold Tier Unlocked!
            </div>
          </div>
          <div
            aria-label="You've unlocked exclusive Gold member benefits including 2X points on every purchase, priority support, and VIP event access."
            class="css-text-146c3p1 r-WebkitBoxOrient-8akbws r-display-krxsd3 r-maxWidth-dnmrzs r-overflow-1udh08x r-textOverflow-1udbk01"
            data-testid="tile-body"
            dir="auto"
            style="color: rgb(51, 51, 51); font-size: 14px;"
          >
            You've unlocked exclusive Gold member benefits including 2X points on every purchase, priority support, and VIP event access.
          </div>
        </div>
      </div>
    </button>
  </button>
</div>
`;
